solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -1919,9 +2051,9 @@ public void removeDocCollectionWatcher(String collection, DocCollectionWatcher w
           if (v == null) return null;
           v.stateWatchers.remove(watcher);
           if (v.canBeRemoved()) {
-            watchedCollectionStates.remove(collection);
             lazyCollectionStates.put(collection, new LazyCollectionRef(collection));
             reconstructState.set(true);
+            CommonTestInjection.injectDelay(); // To unit test race condition
```

Comment:
our common pattern is to hide this behind a java `assert`. See also https://github.com/apache/solr/blob/main/solr/core/src/java/org/apache/solr/query/SolrRangeQuery.java#L183

Reply:
O missed that one. Thanks for catching!